### ISSUE TITLE
PixelData is encoded incorrectly when frame size is not divisible by 8

### ISSUE THREAD

**Issue by fedorov** [QIICR/dcmqi#341]
This issue was discovered while adding corresponding test dataset in #334.

Frame pixel data is initialized in dcmqi here: https://github.com/QIICR/dcmqi/blob/master/libsrc/ImageSEGConverter.cpp#L320-L328.

The frames are then concatenated to form SEG object `PixelData` in `DcmSegmentation::concatFrames` here: https://github.com/commontk/DCMTK/blob/patched-DCMTK-3.6.3_20180205/dcmseg/libsrc/segdoc.cc#L1314.

It appears that something is not quite right with the concatenation. The issue of incorrect pixel data encoding can be easily reproduced with the python code below (this code works with Python 3 and the only extra dependency is pydicom) and [this sample dataset](https://github.com/fedorov/dcmqi/blob/c172992ec4a4036af5fe87bfeb7ed6fb3df381b7/data/segmentations/23x38x3/seg/label.seg) from #334.

```python
import pydicom, sys

d = pydicom.read_file(sys.argv[1])
print(d.Rows)
print(d.Columns)
print(d.NumberOfFrames)
print(d.PixelData)

totalPixels = int(d.Rows*d.Columns*d.NumberOfFrames/8)
if totalPixels%8:
  totalPixels = totalPixels + 1
totalPixels = totalPixels + (totalPixels % 2)
print("Total pixels expected: %i" % totalPixels)

print("Total pixels actual: %i" % len(d.PixelData))


for f in range(d.NumberOfFrames):
  print("Frame %i" % f)
  for i in range(d.Rows):
    for j in range(d.Columns):
      pixelNumber = f*d.Rows*d.Columns+i*d.Columns+j
      byteNumber = int(pixelNumber/8)
      bitPosition = pixelNumber % 8
      value = (d.PixelData[byteNumber] >> bitPosition) & 1
      sys.stdout.write(str(value))
    print("")
```

The output of the code above appears to be identical to the image generated with `dcm2pnm`, as shown in https://github.com/QIICR/dcmqi/pull/334#issuecomment-370449995.

Relevant part of the DICOM standard detailing how bits should be packed in multiframe is in [PS3.5 8.1.1 Pixel Data Encoding of Related Data Elements](http://dicom.nema.org/medical/dicom/current/output/chtml/part05/chapter_8.html#sect_8.1.1). The most important part is there should be no gap/padding between the encoded consecutive frames, and DCMTK implementation is expected to follow that rule, based on earlier communication and debugging.

@michaelonken I have not yet identified the specific problem in the DCMTK concatenation function, but it looks like that is the culprit. I will look into this later next week, but I need to work on something else now and decided to create the issue with the details, in case you want to investigate.

**Comment by michaelonken** [comment]
Looks like my code is causing the problem, so yes, I will investigate this. We had a similar issue before that we discussed (and solved later) at the Hackfest. I will probably not be able to look into this before Friday, though.

**Comment by fedorov** [comment]
@michaelonken any updates?

**Comment by michaelonken** [comment]
@fedorov  I currently look into this and there is something I do not understand, maybe there is an easy explanation:

The input nrrd file has (at least according to its header) 24x38 pixels.
The DICOM segmentation (label.seg), though, has 23x38 pixels (which is also the number of rows/cols/pixels shown by the python program, which is very helpful by the way).

Why is there such a deviation? I do not fully understand the test pipeline run by dcmqi, so not sure at which point (including dcmseg) this deviation takes place. Any ideas?

**Comment by fedorov** [comment]
@michaelonken sorry, I can see how this can be very confusing with the data in a PR... The dataset I used for testing is in this branch: https://github.com/fedorov/dcmqi/tree/add-nonsquare-seg (the same as in this PR: #334).

The specific dataset is under `data/data/segmentations/23x38x3`, and indeed the NRRD file has 23x38 pixels:
```
$ head nrrd/label.nrrd                                                                                                                                                                   2.3.6
NRRD0004
# Complete NRRD file format specification at:
# http://teem.sourceforge.net/nrrd/format.html
type: unsigned short
dimension: 3
space: left-posterior-superior
sizes: 23 38 3
space directions: (0.69999999999999984,0,0) (0,0.69999999999999984,0) (0,0,2.5000000000000004)
kinds: domain domain domain
endian: little
```

To make it easier, I attach the 23x38 dataset here in a zip file.

[23x38x3.zip](https://github.com/QIICR/dcmqi/files/1884196/23x38x3.zip)

By the way, I made the frame dump tool a bit fancier - it now colors the individual pixels according to the odd/even byte, and you can specify the frame to show (it dumps the frame using two approaches - with numpy unpack bits, which is incorrect - just to illustrate the problem in the current implementation in pydicom (the order of bits is reversed, discussed in https://github.com/pydicom/pydicom/pull/293#pullrequestreview-109529983), and using the manual unpacking).

![image](https://user-images.githubusercontent.com/313942/38424682-da0a12b6-397f-11e8-9ee4-9de8a9d10db5.png)

**Comment by fedorov** [comment]
@michaelonken I think I might have found the problem.

I think `shiftRight/Left` operations are not implemented correctly. Since individual bytes are packed from the right (first labeled voxel from the left in the unpacked buffer is the first non-zero bit of the byte from the right in the packed buffer), shifting buffer to the right means individual bytes should be shifted to the left. Here is the proposed patch:

```
diff --git a/dcmseg/libsrc/segdoc.cc b/dcmseg/libsrc/segdoc.cc
index 064142a45..93393e03f 100644
--- a/dcmseg/libsrc/segdoc.cc
+++ b/dcmseg/libsrc/segdoc.cc
@@ -1291,7 +1291,7 @@ void DcmSegmentation::extractFrames(Uint8* pixData,
     }
     // Adapt last byte by masking out unused bits (i.e. those belonging to next frame).
     // A reader should ignore those unused bits anyway.
-    frame->pixData[frame->length-1] = (frame->pixData[frame->length-1] >> (overlapBits)) << (overlapBits);
+    frame->pixData[frame->length-1] = (frame->pixData[frame->length-1] << (overlapBits)) >> (overlapBits);
     // Store frame
     results.push_back(frame);
     // Compute the bitshift created by this frame
@@ -1326,7 +1326,7 @@ void DcmSegmentation::concatFrames(OFVector< DcmIODTypes::Frame* > frames,
     DCMSEG_DEBUG("Packing segmentation frame #" << f+1 << "/" << frames.size());
     // Backup first byte of the destination since it may contain bits of the
     // last frame; mask out those bits not belonging to last frame.
-    firstByte = (writePos[0] >> freeBits) << freeBits;
+    firstByte = (writePos[0] << freeBits) >> freeBits;
     memcpy(writePos, (*frame)->pixData, (*frame)->length);
     // If the last frame left over some unused bits, shift the current frame
     // that number of bits to the right and restore the original bits of the
diff --git a/dcmseg/libsrc/segutils.cc b/dcmseg/libsrc/segutils.cc
index be9e91c4d..6abd33e2c 100644
--- a/dcmseg/libsrc/segutils.cc
+++ b/dcmseg/libsrc/segutils.cc
@@ -139,11 +139,11 @@ void DcmSegUtils::shiftRight(Uint8* buf,
   {
     // Store current byte since we need its last bits
     Uint8 current = buf[x];
-    buf[x] >>= numBits;
+    buf[x] <<= numBits;
     // If there is a carry over from the last handled byte, add it
     buf[x] |= carryOver;
     // Compute carry over bits for next byte handled
-    carryOver = current << (8-numBits); // bits we need to shift over to start (left hand side) of buf[x+1]
+    carryOver = current >> (8-numBits); // bits we need to shift over to start (left hand side) of buf[x+1]
   }
 }

@@ -160,14 +160,14 @@ void DcmSegUtils::shiftLeft(Uint8* buf,
   for (size_t x = 0; x < bufLen-1; x++)
   {
     // Shift current byte
-    buf[x] = buf[x] << numBits;
+    buf[x] = buf[x] >> numBits;
     // isolate portion of next byte that must be shifted into current byte
-    Uint8 next = (buf[x+1] >> (8-numBits));
+    Uint8 next = (buf[x+1] << (8-numBits));
     // Shift current byte
     buf[x] |= next;
   }
   // Shift last byte manually
-  buf[bufLen-1] <<= numBits;
+  buf[bufLen-1] >>= numBits;
 }
```

I tested this change, and the result is that I have `dcm2pnm` output consistent with the input, and the result of conversion back from SEG to NRRD consistent with the source dataset, and consistent with the `dcm2pnm` output.

**Comment by michaelonken** [comment]
Uh, good catch Andrey, thanks for analyzing this further before I could!

Before I take this over I want to make sure to completely understand this, since it's a quite tricky part of the code.

The changes in segdoc.cc (<extract() and concat()) are due to the fact that frames within a single pixel cell (which happens regularly if frame size is not divisible by 8) have been aligned in the wrong order so far. Do you agree?

On the other hand I don't understand why the shiftXXX() operations are implemented the wrong way. Is it not that shifting left by 3 bits should turn  "11111111 00000000" into "11111000 00000000"? Why should any of the bytes/bits move to the right for shiftLeft()? Sorry, I'm sure it's my confusion in this regard, especially since your test shows your correct, too.

**Comment by fedorov** [comment]
> The changes in segdoc.cc (<extract() and concat()) are due to the fact that frames within a single pixel cell (which happens regularly if frame size is not divisible by 8) have been aligned in the wrong order so far. Do you agree?

Not sure if I understood the question correctly. I don't think it is due to mis-alignment within a single pixel cell, but because the shift of all the subsequent frames was not done correctly.

> Is it not that shifting left by 3 bits should turn "11111111 00000000" into "11111000 00000000"? Why should any of the bytes/bits move to the right for shiftLeft()? Sorry, I'm sure it's my confusion in this regard, especially since your test shows you're correct, too.

What you say in your example is correct indeed if the frames are _unpacked_. But they are packed from the right. So for the unpacked buffer:
```
 11111111 00000000 << 3 = 11111000 00000000
```

But since it is packed, the bits destined to be shifted to byte `(i-1)` need to be taken _from the right_ of byte `(i)`, and the packed version of the result above is
```
00011111 00000000
```
Does this make sense?

> Doing this (without applying your changes to the shiftXXX() methods itself), the dcmqi test fails.

What exactly fails? The patch above modifies both the shift functions, but also the `extractFrames()` call. Did you apply the complete patch?

Let me know if it is easier to discuss over phone/screenshare!

> P.S: There is also a test in tutils.cc (dcmtk/dcmseg/tests), that shows how I thought shiftXXX() should work.

Yes, I saw those too. The issue is that they disregard the fact that bytes are packed from the right.

By the way, is there any way to somehow leverage the same code base that `dcm2pnm` uses? Is it really needed to duplicate this functionality?

**Comment by fedorov** [comment]
I agree this issue of bit packing is mind-boggling. If it makes you feel better, a well-known big vendor we all know know had the order wrong on the first try, until David pointed it out to them.

**Comment by fedorov** [comment]
> That being said, my first thought was that based and on top of the changes you applied on segdoc.cc (concat/extract) and the reason behind it, I would need to switch the shift calls in concat/extract from shiftLeft() to shiftRight() and vice versa. Doing this (without applying your changes to the shiftXXX() methods itself), the dcmqi test fails. (I don't know why, but I guess it's my confusion regarding shiftXXX() described above).

I think I understand this better after re-reading it.

No, it is not that you shift calls from shiftLeft() to shiftRight() and vice versa! You need to change the direction of the byte shift within the call.

**Comment by fedorov** [comment]
Thanks to @fladedavid, confirmed that the 23x SEG object generated using the patched code render correctly on Brainlab:

![image](https://user-images.githubusercontent.com/313942/38681208-2c6fe0fc-3e36-11e8-9cc4-97a5e06640c3.png)

[redicomsegrectangulartestdatasets.zip](https://github.com/QIICR/dcmqi/files/1903369/redicomsegrectangulartestdatasets.zip)

**Comment by fedorov** [comment]
Here is the converted SEG that I sent to David Flade earlier.

[fixed.zip](https://github.com/QIICR/dcmqi/files/1903494/fixed.zip)

**Comment by fedorov** [comment]
@michaelonken it's been a while - any updates on the DCMTK side?

**Comment by michaelonken** [comment]
Andrey, i am very busy in the last weeks since a large project is coming to an end. There will be time by the end of last week, sorry that you have to wait, it's totally fine to remind me

**Comment by hmeine** [comment]
In MeVisLab, we cannot load the file, currently.  According to Wolf: 

> The routine `DcmtkDatasetCache::getValue()` in line 54 in file `DCMTree_FileCache_OFFIS.cpp` in
MLDicomTree_OFFIS delivers condition "Invalid Tag" as return of the call 
`OFCondition cond = dElem_->getUncompressedFrame(dSet_, idx, DCMfragment, dest,
bufSize, DCMcolorModel, &dCache_);`
> Why this happens I do not know. Maybe it will be necessary to debug in dcmtk or
to check whether the tag or the tag id the routines works on, is valid. It
looks a bit like a frame or a frame compression which cannot be read or
decompressed from dcmtk. The file itself does not contain any hint about a
compression thus I guess that dcmtk or the dcmtk frame data cache cannot handle
this tag.

We're using DCMTK 3.6.1 20160630, which may be a lot older than what you use.

**Comment by fedorov** [comment]
@hmeine I am not sure what that error is, but perhaps as a separate question - is there a big difficulty DCMTK upgrade for MeVisLab?

**Comment by fedorov** [comment]
@michaelonken any new updates on the estimate for this issue to be fixed in DCMTK?

**Comment by hmeine** [comment]
Updating DCMTK in MeVisLab is possible, in particular after branching for the upcoming 3.1 release. However, it comes with some effort, and I was asked whether I am sure that it will actually fix the problem. Unfortunately, I have not debugged that myself and cannot state more than "I hope so".

**Comment by fedorov** [comment]
Thanks to @michaelonken DCMTK has now been updated with the fix.

**Comment by jcfr** [comment]
@fedorov Should we update the version of DCMTK used in Slicer ?

**Comment by fedorov** [comment]
@jcfr yes, we absolutely should. I am busy with a grant proposal at the moment, and was going to work on that a bit later. If you can easily do it, it would be great!

**Comment by fedorov** [comment]
@jcfr can you please add me to the list of maintainers for this repo so I can push the new branch with the updated DCMTK? https://github.com/commontk/DCMTK

The only extra commit we will need to cherry-pick is this one: https://github.com/commontk/DCMTK/commit/88225bc0a3bcaf2d4e4f212522659a06d80ac3bb

**Comment by fedorov** [comment]
@jcfr thanks and Slicer PR submitted: https://github.com/Slicer/Slicer/pull/976

**Comment by fedorov** [comment]
Confirmed resolved!

**Comment by hmeine** [comment]
I just wanted to give information which I believe is related to this issue:
In MeVisLab, we originally exported SEG files without padding. Then, we got back such files as "invalid" from an important customer who claimed that the rows had to be padded to multiples of 8 bits, and we could not find information in an older DICOM standard version that prohibited that.
Nowadays, [the standard explicitly states that applications shall be prepared to support padded frames since this was not forbidden in older standard versions (Part 5, 8.1.1)](http://dicom.nema.org/dicom/2013/output/chtml/part05/chapter_8.html#sect_8.1.1http://dicom.nema.org/dicom/2013/output/chtml/part05/chapter_8.html#sect_8.1.1).

We now have example files without padding (which seem to be the recommended encoding by today's standard) that make dcmqi crash, apparently because of an "Invalid Tag" error from dcmtk.

**Comment by fedorov** [comment]
Hans, thanks for your note.

I think it would be good to have some better clarification about what your
important customer wanted, and what you actually did.

You first state that your customer "claimed that the rows had to be padded
to multiples of 8 bits", then in your reference to the standard you say
that it "explicitly states that applications shall be prepared to support
padded frames" - those two statements are not the same.

Most importantly, the standard does NOT state that "applications shall be
prepared to support padded frames". What the standard explicitly states is
that (quoting the standard, from the URL you referenced) "some older
applications may send Pixel Data with excess padding, which was not
explicitly prohibited in earlier versions of the Standard".

This statement applies to the padding at the end of the PixelData content,
NOT at a frame or row boundary. It would be quite difficult, if not
impossible, to identify what arbitrary choices were made by the creator
picking and choosing between padding at the row/frame level.

I had a chance to discuss this today with David Clunie, and he confirmed
that the standard indeed talks about the padding at the end of the
PixelData content.

I am not sure what are the actionable items here, since it sounds like you
made your choices for the customer in your application, and the standard is
rather clear in what is the correct and allowable choices.

**Comment by hmeine** [comment]
Hi Andrey, I am sorry for the strange report; in its brevity I think some things were misleading, and I also was not directly involved in the investigation last week, so I may have entirely misunderstood the issue. I just heard "padding" was involved in the communication (with a different partner – the one I confusingly mentioned was long ago) and jumped to conclusions.

So, I have still not entirely understood what's going wrong here, and we're still investigating. I can at least confirm that segimage2itkimage is able to read a multi-segments file that MeVisLab currently has problems with, but we also got single-segment files that segimage2itkimage crashes on, not without first showing `E: Not enough bytes found in Pixel Data element. Found 156 bytes but 158 bytes expected`. So maybe it's not as robust as it could be, but the file may be broken.  (pydicom does not complain about anything, and dciodvfy has a few problems that look unrelated.<sup>[*](#myfootnote1)</sup>)

The segment mask size is:
```
(0028,0008) IS [8]                                      #   2, 1 NumberOfFrames
(0028,0010) US 13                                       #   2, 1 Rows
(0028,0011) US 12                                       #   2, 1 Columns
```
so I would expect 8 * 13 * 12 = 1248 bits.  Without any padding, that would be exactly 156 bytes, but the tool obviously expects 158.

Sorry, this is intermediate info – I will report back when we came to a conclusion w.r.t. the validity of the files.

<a name="myfootnote1">*</a>  I should be explicit about what I consider unrelated problems reported by dciodvfy:
```
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [0] = <-108.85798645019531> - Length invalid for this VR = 19, expected <= 16
Error - Value invalid for this VR - (0x0020,0x0032) DS Image Position (Patient)  DS [1] = <-36.44258117675781> - Length invalid for this VR = 18, expected <= 16
Error - Dicom dataset contains invalid data values for Value Representations
Segmentation
Error - Missing attribute Type 2 Required Element=<PositionReferenceIndicator> Module=<FrameOfReference>
Error - Must start from one, not zero - attribute <DimensionIndexValues>
Error - Must start from one, not zero - attribute <InStackPositionNumber>
```

**Comment by fedorov** [comment]
> we also got single-segment files that segimage2itkimage crashes on, not without first showing `E: Not enough bytes found in Pixel Data element. Found 156 bytes but 158 bytes expected`
> So maybe it's not as robust as it could be, but the file may be broken.
> [...]
>The segment mask size is:
>
>```
>(0028,0008) IS [8]                                      #   2, 1 NumberOfFrames
>(0028,0010) US 13                                       #   2, 1 Rows
>(0028,0011) US 12                                       #   2, 1 Columns
>```
> so I would expect 8 * 13 * 12 = 1248 bits. Without any padding, that would be exactly 156 bytes, but the tool obviously expects 158.

Can you share the file that demonstrates that issue?

> **dcmtk Status** Unfortunately, dcmtk's `getUncompressedFrame()` does not seem to support unpadded frames (which may be yet another reason why Wolf assumed that padding was correct). It seems to be necessary to query the complete PixelData and do the uncompressing yourself. Do you agree with Wolf that this should either be implemented or at least documented?

I am not sure - I have not used that function. Did you ask this on the DCMTK forum? Maybe @michaelonken has any insight on this...

**Comment by hmeine** [comment]
Attached you find a 
[Lesions1.dcm, zipped for GitHub](https://github.com/QIICR/dcmqi/files/5458825/Lesions1.dcm.zip) which contains the described mask. The file is pseudonymized and only contains a lesion mask, so I got permission to share it with you.

**Comment by michaelonken** [comment]
I will look into this. Right now I have no time so it can take a few days...thanks for spending time to isolate this, Hans.

**Comment by fedorov** [comment]
I think there is an issue in DCMTK, and this line should also multiply by the number of frames: https://github.com/DCMTK/dcmtk/blob/master/dcmseg/libsrc/segdoc.cc#L1213.

**Comment by hmeine** [comment]
Yes, that looks like a reasonable explanation, Andrey!

**Comment by michaelonken** [comment]
> I think there is an issue in DCMTK, and this line should also multiply by the number of frames: https://github.com/DCMTK/dcmtk/blob/master/dcmseg/libsrc/segdoc.cc#L1213.

This is only the continuation of the calculation a [few lines above ](https://github.com/DCMTK/dcmtk/blob/f8d0536220789553c509f80078c7dfaef76b5fb4/dcmseg/libsrc/segdoc.cc#L1201) where number of frames is taken into accout.

**Comment by fedorov** [comment]
@michaelonken my point was that the calculation of `remainder` does not take the number of frames into account in https://github.com/DCMTK/dcmtk/blob/master/dcmseg/libsrc/segdoc.cc#L1213, and it is `remainder` that affects whether `bytesRequired` should be bumped.

**Comment by fedorov** [comment]
But I may well be missing something!

**Comment by michaelonken** [comment]
I had time to look at this recently and I agree with you: https://github.com/DCMTK/dcmtk/blob/master/dcmseg/libsrc/segdoc.cc#L1213 is where it happens.

It should read

`remainder = bytesRequired % 8; `

instead since bytesRequired takes into account the number of pixels not for a single frame (old buggy code) but for all frames.

Did you already check whether this does the trick for the files you have at hand, @hmeine?

**Comment by hmeine** [comment]
No, and I do not have my own dcmtk build at hand right now. If you could compile dcmqi against a patched dcmtk, however, you could test running `segimage2itkimage` on the file I attached above: https://github.com/QIICR/dcmqi/files/5458825/Lesions1.dcm.zip

**Comment by AGGDor** [comment]
Was wondering if there was any fix for this yet? I'm having the same issue with a DICOM Segment loaded into slicer "E: Not enough bytes found in Pixel Data element. Found 342 bytes but 344 bytes expected".

**Comment by hmeine** [comment]
I just noticed that we should have opened a separate issue - this recent side track is no longer about incorrect encoding, nor directly about dcmqi, but about dcmtk incorrectly handling a corner case while *decoding* SEG files.

**Comment by AGGDor** [comment]
@hmeine OK, should I open one?

**Comment by fedorov** [comment]
@michaelonken @hmeine I checked the patch suggested in https://github.com/QIICR/dcmqi/issues/341#issuecomment-778253753 on the example from Hans https://github.com/QIICR/dcmqi/files/5458825/Lesions1.dcm.zip, and I do not have an error.

Michael, how do we proceed with fixing the issue in DCMTK? Should I send a PR to the GitHub DCMTK repo?

```
W: PositionReferenceIndicator (0020,1040) absent in FrameOfReferenceModule (type 2)
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
W: ImagePositionPatient (0020,0032) violates maximum VR length in PlanePositionSequence
Row direction: 1 0 0
Col direction: 0 1 0
Z direction: 0 0 1
Total frames: 8
Total frames with unique IPP: 8
Total overlapping frames: 0
Origin: [-108.858, -36.4426, 12.1815]
WARNING: SliceThickness is present and is 3.5. using it!
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
Failed to get CIELab values - initializing to default 43803,26565,37722
```

**Comment by fedorov** [comment]
Ok, I did not realize this has been fixed already in the DCMTK in https://github.com/DCMTK/dcmtk/commit/b665e2ec2d5ce435e28da6c938736dcfa84d0da6!

@AGGDor for this issue to be resolved in Slicer, we will need to wait for the next release of DCMTK (it was fixed after the latest 3.6.6. release), and then update Slicer to use that release.

**Comment by michaelonken** [comment]
yes I recently pushed that into DCMTK master. I am having a report on the same issue from another party, and I am about to check whether matches this one (I think it does), that is why I did not update this thread so far. However, the patch is ok and tested, so the report may be unrelated (or deals with the same problem).

**Comment by fedorov** [comment]
Thank you Michael! Any estimate of when the next DCMTK release should be expected?

**Comment by michaelonken** [comment]
Since we just did a release, I would say at least in half a year. We internally agreed on having one release minimum a year.

**Comment by fedorov** [comment]
Thanks Michael - I will then explore update to 3.6.6 plus patch for this issue.

**Comment by fedorov** [comment]
Need to investigate this again, due to user report related to the ReMIND dataset. See https://support.dcmtk.org/redmine/issues/1132.
